File: Leetcode-Normal/23. Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

## USING DIVIDE AND CONQUER APPROACH.
class Solution:
    def merge2Lists(self, node1, node2):
        head = ListNode()
        dummy = head
        
        while True:
            new_node = None
            if (node1 and node2): ## Both nodes exist.
                if node1.val < node2.val:
                    new_node = ListNode(node1.val)
                    node1 = node1.next
                else:
                    new_node = ListNode(node2.val)
                    node2 = node2.next
            else:
                if node1:
                    new_node = ListNode(node1.val)
                    node1 = node1.next
                if node2:
                    new_node = ListNode(node2.val)
                    node2 = node2.next
            
            dummy.next = new_node ## Assign and Increment.
            dummy = dummy.next
            
            if ((not node1) and (not node2)):
                break
        
        # print("In merge2Lists. newList: ", end=' ')
        # self.printList(head.next)
        
        return head.next
    
    def printList(self, head):
        node = head
        while node:
            print(node.val, end=' ')
            node = node.next
        print("")
    
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if(len(lists) == 0): return None
        
        new_lists = lists ## Keep reference here
        while True: ## Need breaking conditions.
            # print("<WHILE>: ", new_lists)
            if(len(new_lists) == 1): return new_lists[0]
            if(len(new_lists) == 2): return self.merge2Lists(new_lists[0], new_lists[1])
            # new_lists = [self.merge2Lists(new_lists[0], new_lists[1])]

            brand_new_lists = []
            for i in range(0, len(new_lists)):
                if((i%2 == 1)): ## for the even number.
                    merged_head = self.merge2Lists(new_lists[i-1], new_lists[i]) ## merge the two lists.
                    brand_new_lists.append(merged_head)
                else:
                    if (i == (len(new_lists) - 1)): ## only the last element condition checking.
                        brand_new_lists.append(new_lists[-1]) ## append the last element.
            
            ## END OF FOR LOOP
            new_lists = brand_new_lists ## re-assign in the new param.
            
        
        
        return None

File: Leetcode-Premium/RANDOM/23. Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        // return usingPriorityQueue(lists);
        if(lists.length == 0)
            return null;
        
        return divideAndConquer(lists, 0, lists.length - 1);
    }
    
    private ListNode mergeTwo(ListNode node1, ListNode node2){
        ListNode temp = new ListNode(-1);
        ListNode head = temp;

        // ListNode node1 = lists[start];
        // ListNode node2 = lists[end];

        while((node1 != null) && (node2 != null)){
            if(node1.val < node2.val){
                temp.next = new ListNode(node1.val);
                node1 = node1.next;
            }else{
                temp.next = new ListNode(node2.val);
                node2 = node2.next;
            }
            temp = temp.next;
        }

        // any remanining one
        while(node1 != null){
            temp.next = new ListNode(node1.val);
            node1 = node1.next;
            temp = temp.next;
        }

        while(node2 != null){
            temp.next = new ListNode(node2.val);
            node2 = node2.next;
            temp = temp.next;
        }

        return head.next;
    }
    
    private ListNode divideAndConquer(ListNode[] lists, 
                                      int start, int end){
        if((end - start) == 1){
            return mergeTwo(lists[start], lists[end]);
        }else if((end - start) == 0){
            return lists[start];
        }
        else{ // recursive DnC
            int mid = (start + end)/2;
            return mergeTwo(divideAndConquer(lists, start, mid),
                     divideAndConquer(lists, mid+1, end));
        }
    }
    
    private ListNode usingPriorityQueue(ListNode[] lists){
        if(lists.length == 0)
            return null;
        
        
        PriorityQueue<ListNode> pq = new PriorityQueue<>
            ((ListNode n1, ListNode n2) -> (n1.val - n2.val));
        
        ListNode temp = new ListNode(-1);
        ListNode dummy = temp;
        
        for(ListNode node: lists){ // initial node adding
            if(node != null){
                pq.add(node);
            }
        }
        
        while(pq.isEmpty() == false){
            ListNode node = pq.remove();
            if(node != null){
                temp.next = new ListNode(node.val);
                temp = temp.next;
                if(node.next != null)
                    pq.add(node.next); // ADD next node
            }
        }
        
        return dummy.next; // temp was a dummy node
    }
}